add MSBuild binary log (.binlog) component detector

brettfo opened the pull request:
Adds a new NuGet component detector to scan MSBuild binary log (`*.binlog`) files.

There was one major reason to add this: the current NuGet component detector that scans `project.assets.json` isn't always 100% accurate.  This can happen when a given NuGet package has a transitive dependency on a `System.*` package, but the installed SDK has a newer version of that file that is replaced at build time.

One common example: a project has `<PackageReference Include="Microsoft.Extensions.Configuration.Json" Version="6.0.0" />`.  That package has a transitive dependency on `System.Text.Json/6.0.0`.  No matter the SDK used to restore and build the project, the `project.assets.json` file will report that the package `System.Text.Json/6.0.0` was used and implies that it'll be present at runtime.  However, if the `6.0.424` SDK or newer was used to build, the `System.Text.Json.dll` file from the `6.0.0` package is removed and replaced with a newer `System.Text.Json.dll` from the SDK's runtime package.

To avoid reporting a false positive for a package that was replaced at build time (and therefore won't be present at runtime) the MSBuild binary log scanner was added.

If a binary log is generated with the `/bl` flag, every MSBuild event will be recorded.  The new detector in this PR scans each event for the relevant `AddItem` and `RemoveItem` elements (corresponding to things like `<SomeItemGroup Include="..." ... />` and `<SomeItemGroup Remove="..." ... />` and uses that to build a 100% accurate dependency set.

Most of the content of this PR is to support the tests to make them as reliable as possible through 2 different means:

1. Generate a fresh `.binlog` file for each test by running the appropriate `dotnet build ...` command.  This ensures there are no large binary test assets added to this repo; they're simply generated as needed.
2. Mock all necessary NuGet packages to prevent network access.  To accomplish this, 3 common NuGet packages are faked for the tests: `Microsoft.NETCore.App.Ref`, `Microsoft.AspNetCore.App.Ref`, and `Microsoft.WindowsDesktop.App.Ref`.

### Future work

To fully prevent reporting false positives, it would be a good idea to merge the binary log detector and the `project.assets.json` detector so that if a binary log is present and covers a given `.csproj`, _only_ the binary log detector is used and fall back to the original detector otherwise.  Without this work, the binary log detector will properly **not** report `System.Text.Json/6.0.0`, but the `project.assets.json` detector would, so the false positive will still appear.

brettfo committed:
add MSBuild binary log (.binlog) component detector

KirillOsenkov commented on src/Microsoft.ComponentDetection.Detectors/nuget/NuGetMSBuildBinaryLogComponentDetector.cs:
```diff
@@ -0,0 +1,212 @@
+
+            var singleFileComponentRecorder = this.ComponentRecorder.CreateSingleFileComponentRecorder(processRequest.ComponentStream.Location);
+            var buildRoot = BinaryLog.ReadBuild(processRequest.ComponentStream.Stream);
+            this.RecordLockfileVersion(buildRoot.FileFormatVersion);
```
did you mean RecordLogFileVersion?

brettfo replied:
This method comes from the parent class and there's no `Log` variant, only `Lock`, but I figure this would be good enough; it's meant to gather telemetry on what versions of files are encountered, and while the `.binlog` isn't a lock file, it does have the same effect of reporting exactly what ended up in the output directory.

KirillOsenkov commented on src/Microsoft.ComponentDetection.Detectors/nuget/NuGetMSBuildBinaryLogComponentDetector.cs:
```diff
@@ -0,0 +1,212 @@
+    {
+        try
+        {
+            if (!isMSBuildRegistered)
```
can OnFileFoundAsync get called concurrently? If so, you might want a lock here

brettfo replied:
Good catch, I'll push a new commit shortly.

KirillOsenkov commented on src/Microsoft.ComponentDetection.Detectors/nuget/NuGetMSBuildBinaryLogComponentDetector.cs:
```diff
@@ -0,0 +1,212 @@
+                    if (!topLevelDependencies.TryGetValue(projectEvaluation.ProjectFile, out var topLevel))
+                    {
+                        topLevel = new(StringComparer.OrdinalIgnoreCase);
+                        topLevelDependencies[projectEvaluation.ProjectFile] = topLevel;
```
one project file may be evaluated and built several times, you'll often see several evaluations per .csproj. Should we pick the best evaluation somehow or is it fine to do for each evaluation? Restore does an eval, then the build does another eval, and each target framework is a separate eval. We probably want a union of all evaluations?

brettfo replied:
I'll experiment a bit to see if I can find inaccurate results, but generally we _really_ only care about the item groups that NuGet populates like `@(RuntimeCopyLocalItems)` and I think that only happens during the `Restore` evaluation.

brettfo replied:
I just pushed a commit that covers this scenario; check the unit test for some notes, but the end result is that we're scanning the `.binlog` correctly and after a `dotnet build /bl` we'd pick up and report both `Newtonsoft.Json/13.0.3` and `System.Text.Json/8.0.4` and ultimately, that's the end goal: map a `.csproj` (or rather anything that's not `.sln`) to a set of package names and versions that came from it, regardless of the TFM that it came from.

KirillOsenkov commented on src/Microsoft.ComponentDetection.Detectors/nuget/NuGetMSBuildBinaryLogComponentDetector.cs:
```diff
@@ -0,0 +1,212 @@
+                        topLevelDependencies[projectEvaluation.ProjectFile] = topLevel;
+                    }
+
+                    if (doRemoveOperation)
```
a bit worried that you might be getting additem/removeitem from different evaluations in an interleaved way, and since they key by the project path this might get confused.

On the other hand it shouldn't happen because the binlog is a tree, and you're walking the tree linearly, so in theory each evaluation should be processed sequentially one after another.

brettfo replied:
What might be an indication that the traversal is bad?  If I try to process `RemoveItem` but the item isn't already present?  Or is that something that MSBuild won't allow?

KirillOsenkov replied:
yes I guess, but as I said, maybe I'm just being paranoid

KirillOsenkov commented on src/Microsoft.ComponentDetection.Detectors/nuget/NuGetMSBuildBinaryLogComponentDetector.cs:
```diff
@@ -0,0 +1,212 @@
+                                projectDependencies[packageName] = packageVersion;
+                            }
+
+                            project = project.GetNearestParent<Project>();
```
hmm, why are you walking up the project chain? I don't think these items flow to the calling project?

brettfo replied:
I found I needed this to track transitive dependencies.  E.g., if `library.csproj` has `<PackageReference Include="Some.Package" />` and `unitTests.csproj` has `<ProjectReference Include="..\library\library.csproj" />` this will add `Some.Package` first to `library.csproj` and then crawl up the chain to `unitTests.csproj` so that the dependency is properly reported for both projects.  The only oddity with this (and I may have to special case it) is the `.sln` file also appears in a `Project` node, but I _do_ seem to be getting the proper hierarchy.

rainersigwald replied:
I don't think I understand this. `unitTests.csproj` should have `Some.Package` in its assets file and pull assets out of it in its `ResolvePackageAssets`. Can you share a log or project setup where this is necessary?

rainersigwald replied:
I'm saying that the targets that read the assets file should produce `Some.Package` in the transitive case, so walking the project graph here doesn't make sense to me.

brettfo replied:
> ...that read the assets file...

That's the problem, the assets file could be wrong, so I need to crawl it manually.  The PR description explains a scenario where the assets file isn't correct.

ericstj replied:
The assets file will still be a superset of what the project might use.  You get the "real usage" by looking at the outcome of package resolution / conflict resolution.   

If you walk the project graph you end up trying to replay parts of the build.  We shouldn't do that when we can just observe what it did.

Also - the assets file isn't "Incorrect" here - it is correct from NuGet's perspective.  It's just that the build has more policy that it applies to decide if it will actually use a package's contents.  If that package contributes assets which are "older" than some other contribution it will be dropped.  The most common case is framework, as you mentioned -- it's what we designed the feature for in the SDK.  Technically it could be any conflict though - when any two packages try to provide the same file the build will compare them to decide who's copy wins.

KirillOsenkov commented:
As an alternative, you could also see how `binlogtool listnuget msbuild.binlog output.txt` does it. It opens all embedded project.assets.json files, reads the contents into a model, and walks all of them to collect a flat list of all packages and versions used by the build.

I think your approach is fine too (?)

@jeffkl as an expert FYI, in case he wants to take a quick look

KirillOsenkov commented:
actually we could diff the results of binlogtool listnuget with these results, and see if there are any discrepancies. Would be like an oracle!

brettfo commented:
> As an alternative, you could also see how `binlogtool listnuget msbuild.binlog output.txt` does it. It opens all embedded project.assets.json files, reads the contents into a model, and walks all of them to collect a flat list of all packages and versions used by the build.
> 
> I think your approach is fine too (?)
> 
> @jeffkl as an expert FYI, in case he wants to take a quick look

I explicitly don't want the contents of `project.assets.json` because they don't necessarily reflect what ends up in the output.  There are more details in the original description, but in some cases the SDK will replace an assembly that NuGet has already resolved, so according to NuGet, a certain package ended up in the build, but in reality it's not there and shouldn't get reported.

brettfo committed:
add lock around MSBuild registration

brettfo committed:
add test for `.sln` with two unrelated projects

KirillOsenkov commented:
I see, makes sense! 👍🏻

brettfo commented:
For those interested, I ran the package `Microsoft.Extensions.Configuration.Json/6.0.0` through both the `binlogtool` suggegsted above and this PR.  The results are as follows:

`binlogtool`:

```
Microsoft.Extensions.Configuration/6.0.0
Microsoft.Extensions.Configuration.Abstractions/6.0.0
Microsoft.Extensions.Configuration.FileExtensions/6.0.0
Microsoft.Extensions.Configuration.Json/6.0.0
Microsoft.Extensions.FileProviders.Abstractions/6.0.0
Microsoft.Extensions.FileProviders.Physical/6.0.0
Microsoft.Extensions.FileSystemGlobbing/6.0.0
Microsoft.Extensions.Primitives/6.0.0
System.Runtime.CompilerServices.Unsafe/6.0.0
System.Text.Encodings.Web/6.0.0
System.Text.Json/6.0.0

```

This PR:

```
Microsoft.Extensions.Configuration/6.0.0
Microsoft.Extensions.Configuration.Abstractions/6.0.0
Microsoft.Extensions.Configuration.FileExtensions/6.0.0
Microsoft.Extensions.Configuration.Json/6.0.0
Microsoft.Extensions.FileProviders.Abstractions/6.0.0
Microsoft.Extensions.FileProviders.Physical/6.0.0
Microsoft.Extensions.FileSystemGlobbing/6.0.0
Microsoft.Extensions.Primitives/6.0.0
```

The difference being that the SDK (`6.0.425` on my machine) removed the `System.*` packages.

brettfo committed:
don't report dependencies from `.sln` files

rainersigwald commented on src/Microsoft.ComponentDetection.Detectors/nuget/NuGetMSBuildBinaryLogComponentDetector.cs:
```diff
@@ -0,0 +1,224 @@
+        {
+            lock (MSBuildRegistrationGate)
+            {
+                if (!isMSBuildRegistered)
```
nit: you could use `MSBuildLocator.IsRegistered` and not have to carry this around yourself.

rainersigwald commented on src/Microsoft.ComponentDetection.Detectors/nuget/NuGetMSBuildBinaryLogComponentDetector.cs:
```diff
@@ -0,0 +1,224 @@
+
+                    if (doRemoveOperation)
+                    {
+                        topLevel.Remove(packageName);
```
Specifically I think this is going to cause problems if one TF of a project references a thing and the other gets it removed by framework unification.

brettfo replied:
Do you have an example of this?  I'd like to add a test to make sure we don't remove anything that shouldn't be removed.

rainersigwald replied:
Try this:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFrameworks>net472;net8.0</TargetFrameworks>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="System.Text.Json" Version="6.0.0" />
  </ItemGroup>

</Project>
```

This project DOES use STJ 6.0.0, but ONLY for the net472 TF; it should be removed by conflict resolution against the net8.0 framework in that TF.

rainersigwald commented on Directory.Packages.props:
```diff
@@ -21,9 +21,12 @@
     <PackageVersion Include="MSTest.Analyzers" Version="3.5.1" />
     <PackageVersion Include="MSTest.TestAdapter" Version="3.5.1" />
+    <PackageVersion Include="Microsoft.Build.Framework" Version="17.5.0" />
+    <PackageVersion Include="Microsoft.Build.Locator" Version="1.6.1" />
```
I have a concern here. Are you running as a standalone .NET 6 application currently? If so, you won't be able to load MSBuild from new SDKs. Have you tried on a machine that has _only_ .NET 8 installed?

brettfo replied:
This seems to work in the unit tests which should only have .NET6 installed (it's installing [here](https://github.com/microsoft/component-detection/actions/runs/10950179321/workflow?pr=1250#L29) and if I understand that action correctly, it uses `global.json` to determine what to install, so no .NET 8 on the CI machine)

rainersigwald replied:
Right, the problem will be when you run the tool on a different machine which has only .NET 8 installed.

brettfo committed:
report apphost packages from publish

brettfo committed:
use built-in registration check

brettfo committed:
allow the same package with multiple versions

This corresponds to a project with multiple TFMs where the same package is imported in each case, but with a different version each time.

brettfo committed:
report the exact version of the SDK used to build the project

brettfo committed:
track project dependencies by evaluation id

brettfo committed:
introduce types to make the code easier to read

ericstj commented on src/Microsoft.ComponentDetection.Detectors/nuget/NuGetMSBuildBinaryLogComponentDetector.cs:
```diff
@@ -0,0 +1,353 @@
+        // regular restore operations
+        ["NativeCopyLocalItems"] = ("NuGetPackageId", "NuGetPackageVersion"),
+        ["ResourceCopyLocalItems"] = ("NuGetPackageId", "NuGetPackageVersion"),
+        ["RuntimeCopyLocalItems"] = ("NuGetPackageId", "NuGetPackageVersion"),
```
Will you read the value of these late enough in the build to understand that they haven't been reduced by ConflictResolution? 
https://github.com/dotnet/sdk/blob/c3a8f72c3a5491c693ff8e49e7406136a12c3040/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ConflictResolution.targets#L51
https://github.com/dotnet/sdk/blob/c3a8f72c3a5491c693ff8e49e7406136a12c3040/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets#L301-L306
